Add `admerge` as an additional coalescing strategy

dhruvkb opened the pull request:
## Description

This PR
- adds `admerge` as a coalescing strategy which
  - overwrites an existing non-array values (like `merge`)
  - concatenates arrays (like `adjoin`)
- updates documentation to include information about this function

## Examples

```json
{ "a": [1, 2], "b": "value" }
```

```json
{ "a": [3, 4], "b": "new_value" }
```

**adjoin:**
```json
{ "a": [1, 2, 3, 4], "b": "value" }
```

**admerge:**
```json
{ "a": [1, 2, 3, 4], "b": "new_value" }
````

SergioBenitez commented on src/figment.rs:
```diff
@@ -21,10 +21,19 @@ use crate::coalesce::{Coalescible, Order};
+/// an existing value with the same key, while a _joined_ value is discarded if
+/// a value with the same key exists.
+///
+/// | Method                 | Data type | Result              |
```
I like the idea behind this table, but I'm not sure if this is quite the representation that would be most helpful. I would expect exactly four rows, one for each strategy, and for the columns to tell me what each strategy does in a given case. Perhaps the columns should be: strategy (merge, join, adjoin, admerge) linking to each relevant method, array behavior, dict behavior, non-composite value behavior. I'm not sure if "behavior" is the correct word, but we effectively want to say that we keep a previous value, new value, or concatenate, so whatever collectively captures those ideas is the word we want to use for those columns.

dhruvkb replied:
You're right! I've updated the table exactly like you describe in 28c86af.

SergioBenitez commented on src/figment.rs:
````diff
@@ -210,49 +281,8 @@ impl Figment {
-    /// assert_eq!(vec, vec![4, 5]);
-    /// assert_eq!(inner, vec!["world"]);
-    /// ```
+    /// Joins `provider` into the current figment while concatenating vectors.
````
I don't think we should remove the example from this. I know we partly repeat ourselves, but I think that's okay.

dhruvkb replied:
Restored the function-level documentation (with improvements) in 28c86af.

SergioBenitez commented on src/figment.rs:
```diff
@@ -273,6 +303,13 @@ impl Figment {
         self.provide(provider, Order::Merge)
     }
 
+    /// Merges `provider` into the current figment while concatenating vectors.
```
Same thing here. We should have a minimal but representative example, and then delegate the rest to the link.

dhruvkb replied:
Fair point, I added documentation to each individual strategy function in 28c86af.

dhruvkb commented:
@SergioBenitez I updated the documentation in 28c86af.

- Moved all code-related examples to the individual functions. 
- Updated the examples to depict one example of each type (dict, array and non-composite).
- Expanded on the differences of each strategy before the summary table.
- Updated the table columns like you suggested.

SergioBenitez commented:
Merged in https://github.com/SergioBenitez/Figment/commit/174d235a43609bcbbe817c85fb1cd33bca1ca990. Will push a new release shortly.

Thank you for implementing this!

dhruvkb commented:
Thank you for reviewing, and incorporating this. My first Rust contribution 🎉!
